### ISSUE TITLE
Inferred Literals are apparently not promoted to covariants of their own types

### ISSUE THREAD

**Issue by posita** [astral-sh/ty#3143]
### Summary

This comes up when using generics whose types are inferred too narrowly, resulting in errors like: 

    Expected `Iterable[Literal[4]]`, found `list[int]` (invalid-argument-type)

Repro:

```py
from collections.abc import Iterable, Iterator
from typing import TypeVar

_T_co = TypeVar(
    "_T_co",
    covariant=True,  # <-- remove and error goes away
)


class Container(Iterable[_T_co]):
    def __init__(self, init_val: _T_co, /) -> None:
        pass

    def __iter__(self) -> Iterator[_T_co]:
        return iter(())

    def accumulate(self, _other: Iterable[_T_co]) -> "Container[_T_co]":
        return self


Container(4).accumulate(Container(5))  # <-- fails
i: int = 4
j: int = 5
Container(i).accumulate(Container(j))  # <-- still doesn't work, even when i and j have explicit hints
r = list(range(6))
Container(i).accumulate(r)  # <-- left side still too narrow
Container(r[0]).accumulate(Container(5))  # <-- left side okay (works)
```

https://play.ty.dev/ddf3c717-1f7b-440a-a129-971597025311

There's probably a simpler case, but I have yet to figure out how to narrow this one.

This is likely a duplicate of *something*, but I couldn't quite determine exactly what. Possibly fixed with:

* https://github.com/astral-sh/ty/issues/2976 (assuming that fix made it into 0.0.25; found via https://github.com/astral-sh/ty/issues/3007)

Possibly related:

* https://github.com/astral-sh/ty/issues/2280 (no literal promotion with covariant `frozenset`, which sounds like the inverse problem?)
* https://github.com/astral-sh/ty/issues/2884
* https://github.com/astral-sh/ty/issues/3016 (discussion regarding literal promotion relating to contravariance)

### Version

0.0.25

**Comment by AlexWaygood** [comment]
the "still doesn't work, even when `i` and `j` have explicit hints" bit would be solved by #136. After we've implemented that, your `Container(i).accumulate(Container(j))` and `Container(i).accumulate(r)` should no longer cause us to emit errors, because we will understand from the assignment `i: int = 4` that you would like us to upcast `i` from `Literal[4]` to `int` when inferring its type.

The `Container(4).accumulate(Container(5))` call will still cause us to emit an error even after #136 has been implemented, however. We don't generally promote `Literal` types in covariant contexts currently, because we want to infer as precise a type as possible, and it _usually_ does not cause false positives to infer precise `Literal` types in covariant contexts (unlike invariant contexts). This is obviously an exception to that, however.

We could consider promoting `Literal` types in covariant context too, but that might be challenging for us to do in a principled way; it would break some of our special-casing elsewhere for tuples, for example.

**Comment by AlexWaygood** [comment]
Depending on how far we take it, extended full-scope bidirectional inference along the lines of what https://github.com/astral-sh/ty/issues/1473 proposes could also fix the `Container(4).accumulate(Container(5))` call: we could use the context of the argument passed to the method to inform that the `Container(5)` call must be inferred as `Container[Literal[4, 5]]` or `Container[int]` to avoid type errors.

**Comment by posita** [comment]
> We could consider promoting `Literal` types in covariant context too, but that might be challenging for us to do in a principled way; it would break some of our special-casing elsewhere for tuples, for example.

Tuples (and possibly others like frozensets) seem like very special animals deserving of special treatment? 🤔 But I take the point that existing type hinting syntax is likely insufficient to capture those nuances in its own, so having a bunch of logic in the type checker itself seems less than ideal.

> Depending on how far we take it, extended full-scope bidirectional inference along the lines of what [#1473](https://github.com/astral-sh/ty/issues/1473) proposes could also fix the `Container(4).accumulate(Container(5))` call: we could use the context of the argument passed to the method to inform that the `Conatainer(5)` call must be inferred as `Container[Literal[4, 5]]` or `Container[int]` to avoid type errors.

I likely don't know enough about the subject to speak intelligently, but there does seem to be a difference between an explicitly declared `Literal` hint and one inferred by the type checker, especially when examining conformance among inferred literals of precisely the same (not merely covariant) type?

**Comment by AlexWaygood** [comment]
> I likely don't know enough about the subject to speak intelligently, but there does seem to be a difference between an explicitly declared `Literal` hint and one inferred by the type checker, especially when examining conformance among inferred literals of precisely the same (not merely covariant) type?

Indeed! We do already track that -- you can see it in the differences between these two invariant containers:

```py
from typing import Literal, reveal_type

x = 42
reveal_type(x)  # Literal[42]
reveal_type([x])  # list[int]

y: Literal[42] = 42
reveal_type(y)  # Literal[42]
reveal_type([y])  # list[Literal[42]]
```

We refer to the distinction internally as "promotable vs unpromotable" literals (mypy refers to the same distinction as "soft" vs "hard" literal types). But promotable and unpromotable `Literal` types are treated the same way in covariant contexts by ty currently (both are never promoted).

Coming back to tuples again, if we _did_ promote unannotated `Literal` types in covariant contexts, we'd lose the ability to understand this condition as ever evaluating to `Literal[True]` or `Literal[False]`, breaking our understanding of `sys.version_info` branches completely:

```py
import sys

if sys.version_info >= (3, 10):
    ...
```

that could obviously be mitigated by special-casing tuples or by doing our `sys.version_info` handling in a totally different way. But the first feels... undesirable... and the second would be a very big change for us.

**Comment by carljm** [comment]
The core problem with `Container(4).accumulate(Container(5))` is that the `accumulate` method has `_T_co` occurring in contravariant position, which is wrong (since the typevar is declared covariant) and is something that we should emit a diagnostic on. So `Container` is declared as covariant in `_T_co`, but should be invariant, and that discrepancy causes the problem: we feel free to infer a more precise type because it's declared as covariant (which should make it always fine to have a more precise type inferred), but then that more precise type causes a problem because the typevar is used in a contravariant position. So the correct fix here is to mark the typevar as invariant instead of covariant.

So ultimately I think there are two actionable things here: one is #136, and the other is #1017.
